File: server/src/routes/broker/agreements.ts
import { FastifyInstance } from 'fastify';
import prisma from '../../lib/prisma';
import { authMiddleware } from '../../middleware/auth';
import { requireRole } from '../../middleware/rbac';

export default async function agreementRoutes(fastify: FastifyInstance) {
  // List agreements
  fastify.get('/api/broker/agreements', {
    preHandler: [authMiddleware, requireRole('BROKER', 'BROKER_ADMIN', 'INTERNAL')],
    schema: {
      description: 'List all agreements for the broker organisation',
      tags: ['Broker - Agreements'],
      querystring: {
        type: 'object',
        properties: {
          status: { type: 'string', enum: ['DRAFT', 'PROPOSED', 'SIGNED', 'ACTIVE', 'DEFAULTED', 'TERMINATED'] },
          clientId: { type: 'string' },
          page: { type: 'integer', minimum: 1, default: 1 },
          limit: { type: 'integer', minimum: 1, maximum: 100, default: 20 },
        },
      },
    },
  }, async (request, reply) => {
    const { status, clientId, page = 1, limit = 20 } = request.query as any;
    const organisationId = request.auth!.organisationId;

    const where: any = {
      organisationId: request.auth!.role === 'INTERNAL' ? undefined : organisationId,
    };

    if (status) where.status = status;
    if (clientId) where.clientId = clientId;

    const [agreements, total] = await Promise.all([
      prisma.agreement.findMany({
        where,
        skip: (page - 1) * limit,
        take: limit,
        orderBy: { createdAt: 'desc' },
        include: {
          client: true,
          policy: true,
          _count: {
            select: {
              instalments: true,
              payments: true,
            },
          },
        },
      }),
      prisma.agreement.count({ where }),
    ]);

    return {
      data: agreements,
      pagination: {
        page,
        limit,
        total,
        totalPages: Math.ceil(total / limit),
      },
    };
  });

  // Get agreement by ID
  fastify.get('/api/broker/agreements/:id', {
    preHandler: [authMiddleware, requireRole('BROKER', 'BROKER_ADMIN', 'INTERNAL')],
    schema: {
      description: 'Get agreement details by ID',
      tags: ['Broker - Agreements'],
      params: {
        type: 'object',
        properties: {
          id: { type: 'string' },
        },
        required: ['id'],
      },
    },
  }, async (request, reply) => {
    const { id } = request.params as { id: string };
    const organisationId = request.auth!.organisationId;

    const agreement = await prisma.agreement.findFirst({
      where: {
        id,
        organisationId: request.auth!.role === 'INTERNAL' ? undefined : organisationId,
      },
      include: {
        client: true,
        policy: true,
        instalments: {
          orderBy: { sequence: 'asc' },
        },
        payments: {
          orderBy: { collectedAt: 'desc' },
        },
        events: {
          orderBy: { createdAt: 'desc' },
        },
        documents: true,
        commissionLines: true,
      },
    });

    if (!agreement) {
      return reply.code(404).send({ error: 'Agreement not found' });
    }

    return agreement;
  });

  // Create draft agreement
  fastify.post('/api/broker/agreements', {
    preHandler: [authMiddleware, requireRole('BROKER', 'BROKER_ADMIN')],
    schema: {
      description: 'Create a new draft agreement with instalment schedule',
      tags: ['Broker - Agreements'],
      body: {
        type: 'object',
        required: ['clientId', 'policyId', 'principalAmount', 'aprBps', 'termMonths'],
        properties: {
          clientId: { type: 'string' },
          policyId: { type: 'string' },
          principalAmount: { type: 'number' },
          aprBps: { type: 'integer' },
          termMonths: { type: 'integer' },
          startDate: { type: 'string', format: 'date' },
        },
      },
    },
  }, async (request, reply) => {
    const data = request.body as any;
    const organisationId = request.auth!.organisationId;

    // Verify client and policy belong to organisation
    const [client, policy] = await Promise.all([
      prisma.client.findFirst({
        where: { id: data.clientId, organisationId },
      }),
      prisma.policy.findFirst({
        where: { id: data.policyId, organisationId },
      }),
    ]);

    if (!client || !policy) {
      return reply.code(404).send({ error: 'Client or policy not found' });
    }

    // Calculate instalment schedule
    const principalAmount = data.principalAmount;
    const termMonths = data.termMonths;
    const aprBps = data.aprBps;
    const monthlyRate = (aprBps / 10000) / 12;
    const monthlyPayment = principalAmount * monthlyRate / (1 - Math.pow(1 + monthlyRate, -termMonths));

    const startDate = data.startDate ? new Date(data.startDate) : new Date();

    // Create agreement
    const agreement = await prisma.agreement.create({
      data: {
        organisationId,
        clientId: data.clientId,
        policyId: data.policyId,
        principalAmount,
        aprBps,
        termMonths,
        status: 'DRAFT',
        startDate,
        endDate: new Date(startDate.getTime() + termMonths * 30 * 24 * 60 * 60 * 1000),
        outstandingAmount: principalAmount,
      },
    });

    // Create instalments
    const instalments = [];
    for (let i = 0; i < termMonths; i++) {
      const dueDate = new Date(startDate);
      dueDate.setMonth(dueDate.getMonth() + i);
      
      instalments.push({
        agreementId: agreement.id,
        sequence: i + 1,
        dueDate,
        amountDue: Math.round(monthlyPayment * 100) / 100,
        amountPaid: 0,
        status: 'UPCOMING' as const,
      });
    }

    await prisma.instalment.createMany({
      data: instalments,
    });

    // Log event
    await prisma.agreementEvent.create({
      data: {
        agreementId: agreement.id,
        type: 'AGREEMENT_CREATED',
        actorType: request.auth!.role,
        meta: {
          userId: request.auth!.userId,
        },
      },
    });

    // Log audit
    await prisma.auditLog.create({
      data: {
        organisationId,
        actorType: request.auth!.role,
        action: 'CREATE',
        entity: 'AGREEMENT',
        after: agreement,
      },
    });

    return reply.code(201).send(agreement);
  });

  // Propose agreement
  fastify.post('/api/broker/agreements/:id/propose', {
    preHandler: [authMiddleware, requireRole('BROKER', 'BROKER_ADMIN')],
    schema: {
      description: 'Mark agreement as PROPOSED',
      tags: ['Broker - Agreements'],
      params: {
        type: 'object',
        properties: {
          id: { type: 'string' },
        },
        required: ['id'],
      },
    },
  }, async (request, reply) => {
    const { id } = request.params as { id: string };
    const organisationId = request.auth!.organisationId;

    const agreement = await prisma.agreement.findFirst({
      where: {
        id,
        organisationId,
        status: 'DRAFT',
      },
    });

    if (!agreement) {
      return reply.code(404).send({ error: 'Agreement not found or not in DRAFT status' });
    }

    const updated = await prisma.agreement.update({
      where: { id },
      data: { status: 'PROPOSED' },
    });

    // Log event
    await prisma.agreementEvent.create({
      data: {
        agreementId: id,
        type: 'AGREEMENT_PROPOSED',
        actorType: request.auth!.role,
        meta: {
          userId: request.auth!.userId,
        },
      },
    });

    return updated;
  });
}

File: server/routers/agreements.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional
from datetime import datetime, timedelta
from decimal import Decimal
import uuid
from database import get_db
from middleware.auth import AuthContext, get_auth_context
from middleware.rbac import require_role
import models
import schemas
import math

router = APIRouter(prefix="/api/broker/agreements", tags=["Broker - Agreements"])

@router.get("")
async def list_agreements(
    status: Optional[str] = None,
    client_id: Optional[str] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    auth: AuthContext = Depends(get_auth_context)
):
    require_role("BROKER", "BROKER_ADMIN", "INTERNAL")(auth)
    
    query = db.query(models.Agreement)
    
    # Organisation scoping
    if auth.role != "INTERNAL":
        query = query.filter(models.Agreement.organisation_id == auth.organisation_id)
    
    if status:
        query = query.filter(models.Agreement.status == status)
    if client_id:
        query = query.filter(models.Agreement.client_id == client_id)
    
    total = query.count()
    agreements = query.offset((page - 1) * limit).limit(limit).all()
    
    return {
        "data": agreements,
        "pagination": {
            "page": page,
            "limit": limit,
            "total": total,
            "total_pages": math.ceil(total / limit)
        }
    }

@router.get("/{id}")
async def get_agreement(
    id: str,
    db: Session = Depends(get_db),
    auth: AuthContext = Depends(get_auth_context)
):
    require_role("BROKER", "BROKER_ADMIN", "INTERNAL")(auth)
    
    query = db.query(models.Agreement).filter(models.Agreement.id == id)
    
    if auth.role != "INTERNAL":
        query = query.filter(models.Agreement.organisation_id == auth.organisation_id)
    
    agreement = query.first()
    
    if not agreement:
        raise HTTPException(status_code=404, detail="Agreement not found")
    
    return agreement

@router.post("", status_code=201)
async def create_agreement(
    agreement_data: schemas.AgreementCreate,
    db: Session = Depends(get_db),
    auth: AuthContext = Depends(get_auth_context)
):
    require_role("BROKER", "BROKER_ADMIN")(auth)
    
    # Verify client and policy
    client = db.query(models.Client).filter(
        models.Client.id == uuid.UUID(agreement_data.client_id),
        models.Client.organisation_id == uuid.UUID(auth.organisation_id)
    ).first()
    
    policy = db.query(models.Policy).filter(
        models.Policy.id == uuid.UUID(agreement_data.policy_id),
        models.Policy.organisation_id == uuid.UUID(auth.organisation_id)
    ).first()
    
    if not client or not policy:
        raise HTTPException(status_code=404, detail="Client or policy not found")
    
    # Calculate instalment schedule
    principal = agreement_data.principal_amount_pennies / 100  # Convert pennies to pounds
    term_months = agreement_data.term_months
    apr_bps = agreement_data.apr_bps
    monthly_rate = (apr_bps / 10000) / 12
    monthly_payment = principal * monthly_rate / (1 - pow(1 + monthly_rate, -term_months))
    
    signed_at = agreement_data.signed_at or datetime.utcnow()
    activated_at = signed_at + timedelta(days=1)  # Assume activated next day
    
    # Create agreement with explicit timestamps
    now = datetime.utcnow()
    agreement = models.Agreement(
        organisation_id=uuid.UUID(auth.organisation_id),
        client_id=uuid.UUID(agreement_data.client_id),
        policy_id=uuid.UUID(agreement_data.policy_id),
        principal_amount_pennies=agreement_data.principal_amount_pennies,
        apr_bps=agreement_data.apr_bps,
        term_months=agreement_data.term_months,
        broker_fee_bps=agreement_data.broker_fee_bps,
        status=models.AgreementStatusEnum.DRAFT,
        signed_at=signed_at,
        activated_at=activated_at,
        created_at=now,
        updated_at=now
    )
    
    db.add(agreement)
    db.flush()
    
    # Create instalments
    for i in range(term_months):
        due_date = signed_at + timedelta(days=i * 30)
        instalment = models.Instalment(
            agreement_id=agreement.id,
            sequence_number=i + 1,
            due_date=due_date,
            amount_pennies=int(round(monthly_payment * 100)),  # Convert to pennies
            status=models.InstalmentStatusEnum.UPCOMING
        )
        db.add(instalment)
    
    # Note: AgreementEvent model doesn't exist in current database, skipping event creation
    
    # Audit log
    audit_log = models.AuditLog(
        organisation_id=uuid.UUID(auth.organisation_id),
        actor_type=auth.role,
        action="CREATE",
        entity="AGREEMENT",
        after={"id": str(agreement.id)}
    )
    db.add(audit_log)
    
    db.commit()
    db.refresh(agreement)
    
    return agreement

@router.post("/{id}/propose")
async def propose_agreement(
    id: str,
    db: Session = Depends(get_db),
    auth: AuthContext = Depends(get_auth_context)
):
    require_role("BROKER", "BROKER_ADMIN")(auth)
    
    agreement = db.query(models.Agreement).filter(
        models.Agreement.id == id,
        models.Agreement.organisation_id == auth.organisation_id,
        models.Agreement.status == models.AgreementStatusEnum.DRAFT
    ).first()
    
    if not agreement:
        raise HTTPException(
            status_code=404,
            detail="Agreement not found or not in DRAFT status"
        )
    
    agreement.status = models.AgreementStatusEnum.PROPOSED
    
    # Note: AgreementEvent model doesn't exist in current database, skipping event creation
    
    db.commit()
    db.refresh(agreement)
    
    return agreement

@router.delete("/{id}")
async def delete_agreement(
    id: str,
    db: Session = Depends(get_db),
    auth: AuthContext = Depends(get_auth_context)
):
    require_role("BROKER", "BROKER_ADMIN")(auth)
    
    # Find the agreement
    query = db.query(models.Agreement).filter(models.Agreement.id == id)
    
    if auth.role != "INTERNAL":
        query = query.filter(models.Agreement.organisation_id == auth.organisation_id)
    
    agreement = query.first()
    
    if not agreement:
        raise HTTPException(status_code=404, detail="Agreement not found")
    
    # Only allow deletion of draft, proposed, or pending agreements
    if agreement.status not in [models.AgreementStatusEnum.DRAFT, models.AgreementStatusEnum.PROPOSED]:
        raise HTTPException(
            status_code=400,
            detail=f"Cannot delete agreement with status {agreement.status}. Only DRAFT and PROPOSED agreements can be deleted."
        )
    
    # Store agreement data for audit log
    agreement_data = {
        "id": str(agreement.id),
        "client_id": str(agreement.client_id),
        "policy_id": str(agreement.policy_id),
        "status": str(agreement.status),
        "principal_amount_pennies": agreement.principal_amount_pennies
    }
    
    # Delete the agreement (instalments will be cascade deleted)
    db.delete(agreement)
    
    # Audit log
    audit_log = models.AuditLog(
        organisation_id=auth.organisation_id,
        actor_type=auth.role,
        action="DELETE",
        entity="AGREEMENT",
        before=agreement_data
    )
    db.add(audit_log)
    db.commit()
    
    return {"message": "Agreement deleted successfully"}
